### DISCUSSION TITLE
How to choose temporality on metric level?

### DISCUSSION THREAD

**Discussion by StarpTech** [open-telemetry/opentelemetry-go#4921] [General]
Hi, is it possible to configure delta or cumulative temporality per metric? I saw support for it on the Javascript SDK but could not find anything on the go side. AFAIK, I can only specify temporality per exporter.

**Comment by MrAlias** [comment]
You are correct in that the temporality selection is done on the exporter/reader level. A [`TemporalitySelector`](https://pkg.go.dev/go.opentelemetry.io/otel/sdk/metric#TemporalitySelector) is used for this which can select the temporality based on the instrument kind.

There is no way to select temporality more specifically then based on the instrument kind. This is compliance with the specification.

**Reply by StarpTech** [reply]
What's the rationale for this? I read that there are [difference](https://opentelemetry.io/docs/specs/otel/metrics/data-model/#temporality) in costs between delta and cumulative metrics. Those benefits make sense at the metric level.

**Reply by MrAlias** [reply]
Back-ends need to understand how to represent the telemetry they receive. There isn't a known back-end that interprets data temporarily differently based on anything more specific than an instrument type.

**Reply by StarpTech** [reply]
I got but wouldn't it be just an additional flag in the metric export layout similar to `isMonotonic`? Assuming, as an application, I decide to use delta temporality for 2 metrics. The client could filter for the flag to treat them differently.

**Reply by MrAlias** [reply]
Temporality is not a choice for the application instrumentation. It is a concern of the SDK. That is the thing that aggregates data and it does so based on the temporality it is configured it with. How OpenTelemetry as a whole had defined this choice for the SDK is based on how it is implemented in Go.

The modular design does allow you to write your own SDK if you think this should be different though.

**Reply by StarpTech** [reply]
I understand, so it's just the way it is implemented. I thought it might be worth raising this question here. Thanks for clarifying.
